leap_c/utils/torch_autograd.py

```diff
@@ -0,0 +1,77 @@
+"""Utility function to create a PyTorch autograd function"""
+
+import torch
+
+
+def _to_np(data):
+    return tuple(e.detach().cpu().numpy() for e in data)
+
+
+def _to_tensor(data, device, dtype=torch.float32):
+    if isinstance(data, tuple):
+        return tuple(_to_tensor(item, device, dtype) for item in data)
+    return torch.tensor(data, device=device, dtype=dtype)
+
+def _add_none(data):
+    if isinstance(data, tuple):
+        return None, *data
+    return None, data
+
+
+def create_autograd_function(obj) -> type[torch.autograd.Function]:
+    """
+    Create a PyTorch autograd function from an object implementing
+    NumPy-based forward and backward methods.
+
+    An example:
+        class MyAutogradFunction:
+            def forward(
+                self, custom_ctx, *args
+            ) -> np.ndarray | tuple[np.ndarray]:
+                # Perform forward computation using NumPy arrays
+                return outputs
+
+            def backward(
+                self, custom_ctx, *grad_outputs
+            ) -> np.ndarray | tuple[np.ndarray]:
+                # Compute gradients using NumPy arrays
+                return grad_inputs  # A single gradient or a tuple
+
+    Args:
+        obj: An object implementing `forward(custom_ctx, *args)` and
+             `backward(custom_ctx, *grad_outputs)` using NumPy.
+
+    Returns:
+        A callable function that behaves like a PyTorch autograd
+        function.
```

Comment:
maybe "behaves like" is a bit weak, as it actually is an `autograd.Function` via inheritance.